proposals/BVAL-508.md

```diff
@@ -0,0 +1,257 @@
+---
+title: Validating elements contained in a container (like collections)
+layout: default
+author: Emmanuel Bernard
+comments: true
+---
+
+# #{page.title}
+
+[Link to JIRA ticket][jira]  
+[Related JIRA](https://hibernate.atlassian.net/browse/BVAL-499)  
+
+## Problem
+
+It is useful to be able to apply constrains on elements contained in a so called _container_.
+Examples of containers are:
+
+* `Collection`
+* `Optional`
+* JavaFX's `Property`
+
+But today the constraints are applied on the container value itself and not its content.
+
+## Proposition
+
+### Containers
+
+Bean Validation will offer the notion of container of values to validate.
+Containers can contain one value, or several values.
+Extractors for specific containers can be implemented.
+
+TODO: should we differentiate single and plural?
+
+    // Draft proposal for the contract
+    public interface SingleContainerValueExtractor<CONTAINER,CONTAINED> {
+        CONTAINED extractValue(CONTAINER container);
+    }
+    
+    public interface ManyContainerValuesExtractor<CONTAINER,CONTAINED> {
+        Iterable<CONTAINED> extractValue(CONTAINER container);
+    }
+    
+    // Examples
+    
+    //TODO can T remain unspecified? or should it be ? or Object?
+    public class OptionalContainerExtractor<T> implements SingleContainerValueExtractor<Optional<T>,T> {
+        T extractValue(Optional<T> optional) { ... };
+    }
+
+Bean Validation will have out of the box support for containers for `Iterable` and `Map`.
+
+TODO: any other Java SE type to consider?
+TODO: should we support JavaFX Property<T>?
+
+Custom container extractors can be provided via:
+
+* property
```

Comment:
For factory, it's the idea of passing the extractors to the object building the `ValidatorFactory`.